File: typescript/packages/core/src/resources/retrieval.ts
/**
 * Retrieval resource — the agent-native three-step knowledge-graph surface
 * mounted at `/api/v1/query`:
 *
 * - `discover(sandboxId, …)`  → typed menu of relevant context candidates.
 * - `interpret(sandboxId, …)` → 1–2 paragraph brief for one candidate.
 * - `search(sandboxId, …)`    → deep matrix retrieval with a synthesized answer.
 *
 * All three accept either a `window` (a {@link WindowLike} — typically a
 * {@link ContextWindow} from `client.contextWindow.create()`) or a raw
 * `history` array of recent chat turns. When both are set `window` wins.
 * Server caps at 20 turns.
 */

import { BaseResource } from './base.js';
import type { WindowLike } from '../context-window/types.js';

export type ChatRole = 'user' | 'assistant' | 'system';

export interface ChatMessage {
  role: ChatRole;
  content: string;
}

export interface DiscoveryItem {
  id: string;
  score: number;
  /**
   * One-sentence summary. Populated by `copass/copass_1.0` (path string);
   * empty under `copass/copass_2.0` — use `subgraph` instead.
   */
  summary: string;
  /**
   * Tuple of canonical IDs to pin /interpret retrieval to this item.
   *
   * - Under `copass/copass_1.0` this is the full hierarchical path
   *   (root → leaf).
   * - Under `copass/copass_2.0` this is the matched canonical plus
   *   every sub-graph node the query graph identified inside it.
   */
  canonical_ids: string[];
  /**
   * Pre-rendered ASCII tree of the matched canonical's sub-graph,
   * with matched nodes highlighted (⭐) and event timestamps inline.
   * Only populated by the `copass/copass_2.0` preset; absent under
   * `copass/copass_1.0` (which puts a path breadcrumb in `summary`).
   */
  subgraph?: string | null;
  /**
   * Names of the entities in the user's question that this item
   * satisfies. Helps the agent reason about which parts of the
   * question each item answers. Only populated by the
   * `copass/copass_2.0` preset.
   */
  matched_query_nodes?: string[] | null;
}

export interface DiscoverRequest {
  query: string;
  /**
   * Preferred: a {@link WindowLike} (e.g. a {@link ContextWindow}). The
   * resource reads turns via `window.getTurns()` at call time and passes
   * them to the server as `history`.
   */
  window?: WindowLike;
  /** Raw chat turns. Used only when `window` is not supplied. */
  history?: ChatMessage[];
  project_id?: string;
  reference_date?: string;
  /**
   * Retrieval preset selecting the discovery shape. Defaults to
   * `copass/copass_1.0` server-side when omitted. Under
   * `copass/copass_2.0` items carry an additional `subgraph` field
   * with a pre-rendered ASCII tree of the matched canonical, plus a
   * `matched_query_nodes` list of the question entities that resolved
   * to it. The `:thinking` suffix is NOT accepted on `/discover`.
   */
  preset?: SearchPreset;
}

export interface DiscoverResponse {
  /** Markdown title + description orienting the caller on the response shape and how to use it. */
  header: string;
  items: DiscoveryItem[];
  count: number;
  sandbox_id: string;
  project_id?: string;
  query: string;
  /** Short actionable pointer for what to do after picking items. */
  next_steps: string;
}

export interface InterpretRequest {
  query: string;
  /**
   * One or more tuples of canonical IDs to pin interpretation to.
   * Feed the `canonical_ids` list from each DiscoveryItem you want
   * to include.
   */
  items: string[][];
  /** Preferred: a {@link WindowLike}. Wins over `history` when both set. */
  window?: WindowLike;
  /** Raw chat turns. Used only when `window` is not supplied. */
  history?: ChatMessage[];
  project_id?: string;
  reference_date?: string;
  preset?: SearchPreset;
  /** Cap on brief length. Accepts 100–16000; omit for the server default. */
  max_tokens?: number;
}

export interface InterpretCitation {
  canonical_id: string;
  name: string;
  relevance: number;
}

export interface InterpretResponse {
  brief: string;
  citations: InterpretCitation[];
  /** Echo of the `items` tuples the caller sent, for correlation. */
  items: string[][];
  sandbox_id: string;
  project_id?: string;
  query: string;
}

/**
 * Retrieval presets accepted by the Copass API.
 *
 * Canonical names (preferred):
 *   - `copass/copass_1.0` — path-discovery (low-latency default)
 *   - `copass/copass_2.0` — hierarchical-fused per-node embeddings
 *
 * Short aliases `copass/1.0` and `copass/2.0` are also accepted by the
 * server and resolve to the same SearchMatrix; new code should prefer
 * the canonical names.
 *
 * Append `:thinking` to any base preset (e.g. `copass/copass_2.0:thinking`)
 * to run an LLM pre-pass that decomposes the question into sub-questions,
 * executes the base preset on each, and synthesizes one combined answer.
 * The `:thinking` suffix is `/search`-only — `/interpret` and `/discover`
 * reject it.
 */
export type SearchPreset =
  // Canonical names
  | 'copass/copass_1.0'
  | 'copass/copass_2.0'
  | 'copass/copass_1.0:thinking'
  | 'copass/copass_2.0:thinking'
  // Short aliases (kept for backward-compat)
  | 'copass/1.0'
  | 'copass/2.0'
  | 'copass/1.0:thinking'
  | 'copass/2.0:thinking';

export interface SearchRequest {
  query: string;
  /** Preferred: a {@link WindowLike}. Wins over `history` when both set. */
  window?: WindowLike;
  /** Raw chat turns. Used only when `window` is not supplied. */
  history?: ChatMessage[];
  project_id?: string;
  reference_date?: string;
  preset?: SearchPreset;
  detail_level?: 'concise' | 'detailed';
  max_tokens?: number;
}

export interface SearchResponse {
  answer: string;
  preset: SearchPreset;
  execution_time_ms: number;
  warnings?: string[];
  sandbox_id: string;
  project_id?: string;
  query: string;
}

/**
 * Extract `history` from a request: `window.getTurns()` takes precedence,
 * then the explicit `history` array, then an empty array. Strips `window`
 * from the body so it doesn't get serialized.
 */
function resolveBody<T extends { window?: WindowLike; history?: ChatMessage[] }>(
  request: T,
): Omit<T, 'window'> & { history: ChatMessage[] } {
  const { window, history, ...rest } = request;
  const resolved = window ? window.getTurns() : history ?? [];
  return { ...rest, history: resolved } as Omit<T, 'window'> & { history: ChatMessage[] };
}

export class RetrievalResource extends BaseResource {
  discover(sandboxId: string, request: DiscoverRequest): Promise<DiscoverResponse> {
    return this.post<DiscoverResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/discover`,
      resolveBody(request),
    );
  }

  interpret(sandboxId: string, request: InterpretRequest): Promise<InterpretResponse> {
    return this.post<InterpretResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/interpret`,
      resolveBody(request),
    );
  }

  search(sandboxId: string, request: SearchRequest): Promise<SearchResponse> {
    return this.post<SearchResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/search`,
      resolveBody(request),
    );
  }
}

File: python/copass-core/src/copass_core/resources/retrieval.py
"""Retrieval resource — ``discover`` / ``interpret`` / ``search``.

Hand-ported from ``typescript/packages/core/src/resources/retrieval.ts``.

All three accept either a ``window`` (a :class:`WindowLike` — typically
a ``ContextWindow``) or a raw ``history`` list. When ``window`` is set
it wins and ``history`` is ignored. Server caps at 20 turns.
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Any, Dict, List, Optional

from copass_core.resources.base import BaseResource
from copass_core.types import ChatMessage, SearchPreset, WindowLike


def _turns_from(window: Optional[WindowLike]) -> List[Dict[str, str]]:
    if window is None:
        return []
    turns = window.get_turns()
    if not turns:
        return []
    return [_turn_to_dict(t) for t in turns]


def _turn_to_dict(turn: Any) -> Dict[str, str]:
    """Accept either a :class:`ChatMessage` dataclass or a plain dict
    (``{"role": ..., "content": ...}``) so callers can mix and match."""
    if isinstance(turn, ChatMessage):
        return {"role": turn.role, "content": turn.content}
    if isinstance(turn, dict) and "role" in turn and "content" in turn:
        return {"role": str(turn["role"]), "content": str(turn["content"])}
    raise ValueError(
        f"Unexpected turn type {type(turn).__name__}; expected ChatMessage or dict."
    )


def _history(
    window: Optional[WindowLike],
    history: Optional[List[Any]],
) -> List[Dict[str, str]]:
    """``window`` wins; fall back to ``history``; empty list if both None."""
    if window is not None:
        return _turns_from(window)
    if history:
        return [_turn_to_dict(t) for t in history]
    return []


class RetrievalResource(BaseResource):
    """``/api/v1/query/sandboxes/{sandbox_id}/{discover,interpret,search}``."""

    async def discover(
        self,
        sandbox_id: str,
        *,
        query: str,
        window: Optional[WindowLike] = None,
        history: Optional[List[Any]] = None,
        project_id: Optional[str] = None,
        reference_date: Optional[str] = None,
        preset: Optional[SearchPreset] = None,
    ) -> Dict[str, Any]:
        """Ranked menu of context items relevant to ``query``.

        Window-aware: repeated calls skip items already surfaced
        earlier in the conversation.

        ``preset`` selects the discovery shape (defaults to
        ``copass/copass_1.0`` server-side when omitted). Under
        ``copass/copass_2.0`` each returned item carries an additional
        ``subgraph`` field with a pre-rendered ASCII tree of the matched
        canonical's sub-graph, plus a ``matched_query_nodes`` list of
        the question entities that resolved to it. The ``:thinking``
        suffix is NOT accepted on ``discover``.
        """
        body: Dict[str, Any] = {
            "query": query,
            "history": _history(window, history),
        }
        if project_id is not None:
            body["project_id"] = project_id
        if reference_date is not None:
            body["reference_date"] = reference_date
        if preset is not None:
            body["preset"] = preset
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/discover",
            body,
        )

    async def interpret(
        self,
        sandbox_id: str,
        *,
        query: str,
        items: List[List[str]],
        window: Optional[WindowLike] = None,
        history: Optional[List[Any]] = None,
        project_id: Optional[str] = None,
        reference_date: Optional[str] = None,
        preset: Optional[SearchPreset] = None,
        max_tokens: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Synthesized 1–2 paragraph brief pinned to specific
        ``items`` picked from a prior ``discover`` call."""
        body: Dict[str, Any] = {
            "query": query,
            "items": items,
            "history": _history(window, history),
        }
        if project_id is not None:
            body["project_id"] = project_id
        if reference_date is not None:
            body["reference_date"] = reference_date
        if preset is not None:
            body["preset"] = preset
        if max_tokens is not None:
            body["max_tokens"] = max_tokens
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/interpret",
            body,
        )

    async def search(
        self,
        sandbox_id: str,
        *,
        query: str,
        window: Optional[WindowLike] = None,
        history: Optional[List[Any]] = None,
        project_id: Optional[str] = None,
        reference_date: Optional[str] = None,
        preset: Optional[SearchPreset] = None,
        detail_level: Optional[str] = None,
        max_tokens: Optional[int] = None,
    ) -> Dict[str, Any]:
        """One-shot synthesized natural-language answer."""
        body: Dict[str, Any] = {
            "query": query,
            "history": _history(window, history),
        }
        if project_id is not None:
            body["project_id"] = project_id
        if reference_date is not None:
            body["reference_date"] = reference_date
        if preset is not None:
            body["preset"] = preset
        if detail_level is not None:
            body["detail_level"] = detail_level
        if max_tokens is not None:
            body["max_tokens"] = max_tokens
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/search",
            body,
        )


__all__ = ["RetrievalResource"]
